syntax = "proto3";


import "google/protobuf/timestamp.proto";
import "common.proto";

package market_data;

service MarketDataService {
    rpc GetOrderBook (GetSymbol) returns (stream OrderBook);
    rpc GetTrades(GetSymbol) returns (stream Trade);
    rpc GetUnderlier(GetSymbol) returns (stream Underlier);
    rpc GetFundingRate(GetSymbol) returns (stream FundingRate);

    rpc GetRefData(common.Empty) returns (RefData);
    rpc GetBbo(common.Empty) returns (stream OrderBook);
    rpc GetLatestCandle(GetCandle) returns (stream Candle);
}


message GetSymbol {
    string symbol = 1;
}

message GetCandle {
    string symbol = 1;
    common.CandlesInterval interval = 2;
}

message OrderBookEntry {
    double price = 1;
    double quantity = 2;
}

message OrderBook {
    string symbol = 1;
    google.protobuf.Timestamp timestamp = 2;
    repeated OrderBookEntry bids = 3;
    repeated OrderBookEntry asks = 4;
    int64 sequence = 5;
}

message Trade {
    string trade_id = 1;
    string symbol = 2;
    double price = 3;
    double quantity = 4;
    google.protobuf.Timestamp trade_timestamp = 5;

    common.OrderDirection aggressor_side = 6;
}

message Underlier {
    string symbol = 1;
    double price = 2;
    
    google.protobuf.Timestamp timestamp = 99;
}

message FundingRate {
    string symbol = 1;
    double rate = 2;
    uint64 time_remaining_seconds = 3;

    google.protobuf.Timestamp timestamp = 99;
}

message DayMarketHours {
    string open = 1;
    string close = 2;
}

message MarketHours {
    string tz = 1;
    DayMarketHours monday = 2;
    DayMarketHours tuesday = 3;
    DayMarketHours wednesday = 4;
    DayMarketHours thursday = 5;
    DayMarketHours friday = 6;
    DayMarketHours saturday = 7;
    DayMarketHours sunday = 8;
}


message SymbolRefData {
    string clobPairId = 1;
    string symbol = 2;
    common.SymbolStatus status = 3;
    double underlier_price = 4;
    double price_change_24h = 5;
    double volume_24h = 6;
    double trades_24h = 7;
    double next_funding_rate = 8;
    string tick_size = 11;
    string lot_size = 12;
    string min_price = 13;
    string max_price = 14;
    string min_quantity = 15;
    string max_quantity = 16;

    string base_asset = 17;
    string quote_asset = 18;
    string margin_asset = 19;
    repeated common.OrderTimeInForce order_time_in_force = 20;
    repeated common.OrderType order_types = 21;
    MarketHours market_hours = 22;  // New field for market hours

}

message RefData {
    repeated SymbolRefData symbols = 1;
    google.protobuf.Timestamp timestamp = 99;
}

message Candle {
    string symbol = 1;
    google.protobuf.Timestamp timestamp = 2;
    double open = 3;
    double high = 4;
    double low = 5;
    double close = 6;
    double volume = 7;
    int64 period = 8; // in seconds
}
